### ISSUE TITLE
Token string to numerical (float) in Jax.

### ISSUE THREAD

**Issue by ScottSauers** [google/jax#20452]
### Description

I am still working on a minimal reproducible example. I seem to not be able to replicate an exact error in a short file because the error changes based on implementation details. Attempting to separate this step from the model is difficult, because it uses the model's output.

Here is what I am attempting to do:

Take in an array like <ShapedArray(float32[1,17])>. In my code, this represents logits, of shape (1, 17).

I'd like to use softmax on these logits to get token indices, and then decode these indices into their textual, human-readable representation. This representation will be a number (a float or int), which I want to use in Jax later.

Is this possible at all?

### System info (python version, jaxlib version, accelerator, etc.)

 Python 3.11. MacOS. Apple Silicon / plain CPU.

jax:    0.4.25
jaxlib: 0.4.25
numpy:  1.26.4
python: 3.11.6 (main, Oct  2 2023, 20:46:17) [Clang 14.0.0 (clang-1400.0.29.202)]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Darwin', node='comp-7.local', release='23.2.0', version='Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:34 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T8103', machine='arm64')

**Comment by selamw1** [comment]
Hi @ScottSauers 
Your process for converting model logits to a textual, numerical representation is feasible and aligns well with common practices in natural language processing. This can be achieved using Jax.
Let us know if the following concise guide helps you:

1. **Softmax on Logits**: Apply [`jax.nn.softmax`](https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.softmax.html#jax.nn.softmax) to convert logits of shape (1, 17) into probabilities.

2. **Extract Token Indices**: Use [`jnp.argmax`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.argmax.html#jax.numpy.argmax)  to select the token with the highest probability. For sampling with randomness, consider different sampling strategies.

3. **Decode to Textual Representation**: Utilize your model's tokenizer to decode the selected token index into its human-readable text format.

4. **Convert Text to Number**: Convert the textual representation to a numerical value (`int` or `float`) using `int()` or `float()`.

Sample example implementation (adjust according to your specific tokenizer and model output):

```Python
import jax.numpy as jnp
from jax.nn import softmax
import numpy as np

# Sample logits 
logits = jnp.array([[0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0,
                    1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7]])

# Convert logits to probabilities using softmax
probabilities = softmax(logits)

# Find the index of the most probable token
token_index = jnp.argmax(probabilities).item()  

# Assuming you have a decoder function for token indices
def my_decoder(index):
  # Implement your actual decoding logic here
  # This example assumes a simple mapping (not a real use case)
  return str(index)

# Decode the token index
text_representation = my_decoder(token_index)

# Optionally convert text to a number (if applicable)
number_representation = None
try:
  # Attempt conversion based on your specific use case
  number_representation = float(text_representation)
except ValueError:
  print("Warning: Could not convert textual representation to a number.")

print("Token Index:", token_index)
print("Textual Representation:", text_representation)
if number_representation is not None:
  print("Number Representation:", number_representation)

```

This code is a general guide and will need adjustments based on your actual tokenizer and how your model is structured. Specifically, the way you decode token indices to their textual representation will depend on the tokenizer or decoding mechanism you have in place.

**Comment by ScottSauers** [comment]
Are you sure this is possible? This is pretty close to my set up which I tried for a while and it didn't work.

**Comment by jakevdp** [comment]
It's difficult to tell what the problem is here from just a verbal description of the code. Could you put together a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) showing what you tried to do, and how it failed?

**Comment by jakevdp** [comment]
Closing due to lack of activity. Feel free to ask another question if you're still having this issue!
